### PR TITLE
Added tests for collections, promises and function length

### PR DESCRIPTION
**Pull request by shvaikalesh**
1. `@@species` are missing in `WeakMap` and `WeakSet` constructors. Added the test to check for that: c48968a.
2. Some implementations (V8 prior to 4.3 and some polyfills) do not support iterables in `Promise.all` and `Promise.race`. Added tests to check that: 75bb18b.
3. In ES6 version of specification, contrary to ES5, `length` property of `function` instances is `configurable`. Added test for that: 091df00.
4. In early versions of ES6 spec weak collections prototypes had `clear` methods that were later removed. Added tests for their absence: f08d8f5.
5. Firefox 36 throws on `null` in constructors (5c1d6f3) and does not invoke monkey-patched `set` and `add` methods (56ee56c).

### CODE REVIEW

File: data-es6.js [hunk 1]
```diff
@@ -3586,6 +3586,54 @@ exports.tests = [
         iojs:        true,
       },
     },
+    'constructor accepts null': {
```
Comment (ljharb) [inline data-es6.js]: Please add `es6shim: true` to this test, since the `es6-shim`'s `Map` constructor does accept null.

File: data-es6.js [hunk 2]
```diff
@@ -3586,6 +3586,54 @@ exports.tests = [
...
+        firefox38:   true,
+        firefox39:   true,
+        firefox40:   true,
+        chrome43:    true,
+        chrome45:    true,
+        webkit:      true,
+        iojs:        true,
+      },
+    },
+    'constructor invokes set': {
```
Comment (ljharb) [inline data-es6.js:3611]: please add `es6shim: true` to this test as well.

File: data-es6.js [hunk 3]
```diff
@@ -3893,6 +3941,54 @@ exports.tests = [
         iojs:        true,
       },
     },
+    'constructor accepts null': {
```
Comment (ljharb) [inline data-es6.js]: Please add `es6shim: true` to this test, since the `es6-shim`'s `Set` constructor does accept null

File: data-es6.js [hunk 4]
```diff
@@ -3893,6 +3941,54 @@ exports.tests = [
...
+        firefox38:   true,
+        firefox39:   true,
+        firefox40:   true,
+        chrome43:    true,
+        chrome45:    true,
+        webkit:      true,
+        iojs:        true,
+      },
+    },
+    'constructor invokes add': {
```
Comment (ljharb) [inline data-es6.js:3967]: please add `es6shim: true` to this test as well.

File: data-es6.js [hunk 5]
```diff
@@ -4275,13 +4419,31 @@ exports.tests = [
         iojs:        true,
       },
     },
-    'WeakMap[Symbol.species]': {
+    'WeakMap.prototype.clear is missing': {
       exec: function () {/*
-        var prop = Object.getOwnPropertyDescriptor(WeakMap, Symbol.species);
-        return 'get' in prop && WeakMap[Symbol.species] === WeakMap;
+        return !("clear" in WeakMap.prototype);
```
Comment (ljharb) [inline data-es6.js]: I feel like this test should only return "false" if both `WeakMap` is present, and `WeakMap#clear` is present - in any other case, this security hole doesn't exist. (Same with the similar `WeakSet` tests)

File: data-es6.js [hunk 6]
```diff
@@ -5668,6 +5896,35 @@ exports.tests = [
         iojs:        true,
       },
     },
+    'Promise.all supports iterables': {
```
Comment (ljharb) [inline data-es6.js]: Please add `es6shim: true` to this test.

File: data-es6.js [hunk 7]
```diff
@@ -5701,6 +5958,35 @@ exports.tests = [
         iojs:        true,
       },
     },
+    'Promise.race supports iterables': {
```
Comment (ljharb) [inline data-es6.js]: Please add `es6shim: true` to this test.

File: data-es6.js [hunk 8]
```diff
@@ -5701,6 +5963,36 @@ exports.tests = [
...
+      exec: function () {/*
+        var fulfills = Promise.race(new Set([
```
Comment (zloirock) [inline data-es6.js:5968]: `Promise` methods do not depend on `Set`. For  generic iterables  available `__createIterableObject`.

**Comment by webbedspace** [comment]
Previously, we haven't tested for the _absence_ of ES draft methods/properties before, so points 1 and 4 are somewhat extending the compat table's current scope. Moreover, it seems to suggest dozens of other absence tests of dubious relevance: should we check for the absence of `"create" in Symbol`, `"toMethod" in Function.prototype`, or `"Realm" in Reflect`? I'm not really a fan of this idea at present.

**Comment by ljharb** [comment]
@webbedspace Essentially, I'd say it depends on what the thing is testing. If we're testing for a method's presence and compliance, then absence fails it. But, if we're testing for something _negative_ - specifically, known violations and/or bugs - I'd say absence passes it, because the environment lacks the problem (which is all the test is ascertaining).

**Comment by webbedspace** [comment]
There are actually a few tests which check for absences of properties which are explicit prohibitions (such as `prototype` on arrow functions), but I refer in this case to unspecified properties - extensions to the built-ins beyond those in the spec are [allowed except where noted](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-error-handling-and-language-extensions), so even draft methods like `Function#toMethod` are permitted in a conforming ES6 implementation.

**Comment by ljharb** [comment]
In this case, `WeakMap#clear` is not a permitted extension, because a `WeakMap` _must_ not allow any iteration or mutation except when you already have the key. It's a security invariant.

**Comment by webbedspace** [comment]
Well:
- I've just read through [the WeakMap definition again](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-weakmap-objects) and it seems such an invariant against removal of key-value pairs isn't explicitly specified, at least not in the style of other invariants such as [those for proxies](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-invariants-of-the-essential-internal-methods).
- This pull request's test only checks for the absence of a property called "clear", without examining the behaviour of any present "clear" property. If a present `WeakMap#clear` method did nothing when invoked, it would be compliant.

**Comment by ljharb** [comment]
@webbedspace The entire motivation for removing `WeakMap#clear` from the spec was the security invariant. If it's not clearly specified enough I can get a note added to the next edition - suffice it to say, it's prohibited. In this case, the only engines that have a property called "clear" _do_ in fact clear the WeakMap. None exist that have a noop "clear" method.

You could certainly write a test that, if the "clear" method is present, attempts to clear the map, and fails if it is indeed cleared. However, I think that'd be a waste of time until such time as an engine exists that does this.

**Comment by webbedspace** [comment]
I'll write that test anyway.

I'll accept this pullreq now and make all the suggested changes myself.

**Comment by ljharb** [comment]
Sounds good to me

**Comment by zloirock** [comment]
WTF? Merged? `Promise` tests uses `Set`, `es6-shim` `WeakSet` 2/9, `WeakMap` 1/10?

**Comment by ljharb** [comment]
I agree the `Promise` tests shouldn't use `Set`, that was a good call.

But yes, anything that doesn't have `WeakMap#clear` should pass that test, even if it doesn't have `WeakMap`. Same with `WeakSet`. I can make that even more certain by actively deleting those methods when they're detected.

**Comment by webbedspace** [comment]
I disagree. The negative test (absence of `clear`) is designed to check if the existing feature (WeakMap) adheres to the security invariant (by not exposing a given API). If there is no existing feature, then there is no adherence. But, I recall we've been down a similar rhetorical path before...
